com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs

```diff
@@ -504,6 +509,196 @@ internal void InvokeBehaviourOnGainedOwnership()
+        {
+            for (int i = 0; i < ChildNetworkBehaviours.Count; i++)
+            {
+                ChildNetworkBehaviours[i].OnNetworkObjectParentChanged(parentNetworkObject);
+            }
+        }
+
+        internal void SetCachedParent(Transform parentTransform)
+        {
+            m_CachedParent = parentTransform;
+        }
+
+        private void OnTransformParentChanged()
+        {
+            if (transform.parent == m_CachedParent)
+            {
+                return;
+            }
+
+            if (NetworkManager == null || !NetworkManager.IsListening)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new NotListeningException($"{nameof(NetworkManager)} is not listening, start a server or host before reparenting"));
+                return;
+            }
+
+            if (!NetworkManager.IsServer)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new NotServerException($"Only the server can reparent {nameof(NetworkObject)}s"));
+                return;
+            }
+
+            if (!IsSpawned)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new SpawnStateException($"{nameof(NetworkObject)} can only be reparented after being spawned"));
+                return;
+            }
+
+            var parentTransform = transform.parent;
+            if (parentTransform != null)
+            {
+                var parentObject = transform.parent.GetComponent<NetworkObject>();
+                if (parentObject != null)
+                {
+                    m_LatestParent = parentObject.NetworkObjectId;
+                }
+                else
+                {
+                    if (NetworkLog.CurrentLogLevel <= LogLevel.Error)
+                    {
+                        NetworkLog.LogError($"Invalid parenting, {nameof(NetworkObject)} moved under a non-{nameof(NetworkObject)} parent, will fix and move to the scene root");
+                    }
+
+                    m_LatestParent = null;
+                }
+            }
+            else
+            {
+                m_LatestParent = null;
+            }
+
+            using (var buffer = PooledNetworkBuffer.Get())
+            {
+                using (var writer = PooledNetworkWriter.Get(buffer))
+                {
+                    writer.WriteUInt64Packed(NetworkObjectId);
+                    WriteNetworkParenting(writer, m_IsReparented, m_LatestParent);
+                }
+
+                for (int i = 0; i < NetworkManager.ConnectedClientsList.Count; i++)
+                {
+                    var targetClientId = NetworkManager.ConnectedClientsList[i].ClientId;
+                    if (Observers.Contains(targetClientId))
+                    {
+                        NetworkManager.MessageSender.Send(targetClientId, NetworkConstants.PARENT_SYNC, NetworkChannel.Internal, buffer);
```

Comment:
out of curiosity, why aren't we using netvars to track this?
Custom messages makes this harder to track in any eventual AoI / Prediction / Server side rewind, no?

Reply:
I don't think NetVar is the right tool for this job.

Reply:
Any reason why not?

Reply:
> Custom messages makes this harder to track in any eventual AoI / Prediction / Server side rewind, no?
what you see inline above is not a "custom message", custom messages are something completely different.

> Any reason why not?
I'd argue why so?

NetVar is being checked every-frame, does some other things and will do some other things (that I don't fully understand yet, maybe I don't have to?) which is not needed here because we precisely control when parent changes and know exactly when it happens.
`NetworkObject` parenting could live without NetVar coupling and I believe de-coupling is a good thing.

Reply:
I need to look more closely but I don't want to forget this thought...so with this change, is the NetworkTransform always sending the relative-to-parent or global position as before?

Regardless, my intuition is suggesting that I suggest this - what if we atomically sent either just a position change (if the parent hasn't changed, as today) *or* parent change plus position/rot change.  That way you never have the weird case where the parent changed with a mis-matched transform position causing jumping, etc.  This parenting only applies to transform (yes?) so it doesn't seem like a bolt on.

Reply:
things could've done better for sure but things are being done this way due to our current MLAPI architecture (NetworkObject-NetworkBehaviour relationship, their relation to scene hierarchy and NetworkTransform doing transform-related stuff. I think both of you know where I stand in terms of NO, NB & NT — NO & NB should not even form any transform-based hierarchy at all, if any at all).

apart from that, we can still do things in `NetworkTransform` realm. we do have a `OnNetworkObjectParentChanged()` callback when `NetworkObject` reparenting happens (on both server and client sides). that could be used to switch between local vs world relative positioning or any other custom optimization at `NetworkTransform`-level.

I added a placeholder method implementation into `NetworkTransform.cs` where I put a comment saying:

> // TODO: handle parent NetworkObject change and potentially optimize pos/rot/scale replication?

that'd hopefully guide us later in the future (if and) when we decide to discover optimization possibilities on `NetworkTransform`.